text/3703-wrapping-from.md

```diff
@@ -0,0 +1,277 @@
+
+So there is, in fact, a coherent wrapping arithmetic on `NonZero` itself.
+
+It would thus have
+```rust
+assert_eq!(<NonZero<u8>>::wrapping_from(NonZero(  1_u16)), NonZero(  1_u8));
+assert_eq!(<NonZero<u8>>::wrapping_from(NonZero(  2_u16)), NonZero(  2_u8));
+assert_eq!(<NonZero<u8>>::wrapping_from(NonZero(254_u16)), NonZero(254_u8));
+assert_eq!(<NonZero<u8>>::wrapping_from(NonZero(255_u16)), NonZero(255_u8));
+assert_eq!(<NonZero<u8>>::wrapping_from(NonZero(256_u16)), NonZero(  1_u8));
+assert_eq!(<NonZero<u8>>::wrapping_from(NonZero(257_u16)), NonZero(  2_u8));
+assert_eq!(<NonZero<u8>>::wrapping_from(NonZero(509_u16)), NonZero(254_u8));
+assert_eq!(<NonZero<u8>>::wrapping_from(NonZero(510_u16)), NonZero(255_u8));
+assert_eq!(<NonZero<u8>>::wrapping_from(NonZero(511_u16)), NonZero(  1_u8));
+assert_eq!(<NonZero<u8>>::wrapping_from(NonZero(512_u16)), NonZero(  2_u8));
+```
+and analogously for other types.
+
+This would make sense if we decided to add a `wrapping_add` method to `NonZero`
+such that `NonZero(255_u8).wrapping_add(1)` → `NonZero(1_u8)`.
+
+But it's not clear that we *want* to do that.  It might be that we want people
+to think about `NonZero` as behaving like the underlying type, just with a value
+restriction, rather than a different cycle of values.
+
+As such, this RFC doesn't propose this directly, just like how we don't have
+`NonZero<u32>::wrapping_add` even as unstable.  So far we only have methods on
+`NonZero` which don't need to resolve that question: `saturating_add` and
+`checked_add` for *unsigned* types only, which by only being able to
+strictly-increase a value work the same as on the underlying type.
+
+An implementation of `wrapping_from` would need to deal with that issue, so this
+RFC leaves it as something to consider in the future.
+
+
+# Prior art
+[prior-art]: #prior-art
+
+This is obviously inspired by the `From` and `TryFrom` traits.  (The latter being why there's no `num::CheckedFrom` needed.)
+
+Much discussion related to the general idea happened in [RFC PR 2484](https://github.com/rust-lang/rfcs/pull/2484).
+
+The `num_traits` crate has a [`FromPrimitive`](https://docs.rs/num-traits/latest/num_traits/cast/trait.FromPrimitive.html) trait.
+
+C++ has a `static_cast` primitive which, like Rust's `as`, can do a great many things.  Boost added a [`numeric_cast`]
+function to be more restrictive, though it's more like `T::try_from(x).unwrap()` than like the `wrapping_from` proposed here.
+
+[`numeric_cast`]: https://live.boost.org/libs/numeric/conversion/doc/html/boost_numericconversion/improved_numeric_cast__.html
+
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+(None yet known.)
+
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+Once the trait exists, there's many possible different ways to expose it.
+
+There could, for example, be some sort of `.truncate()` method similar to `.into()`,
+perhaps tweaked for better turbofishing.
```

Comment:
Presumably, this would be via a `WrappingInto` trait, not `truncate`, and be `wrapping_into`. I think that it's fine to exclude from the initial RFC, but this should probably be at least sort-of fleshed out here.